exts/virtio_blk/lib.rs

```diff
@@ -0,0 +1,228 @@
+#![no_std]
+
+extern crate alloc;
+#[macro_use]
+extern crate kerla_api;
+
+use core::mem::size_of;
+
+use kerla_api::address::VAddr;
+use kerla_api::driver::{DeviceProber, Driver};
+use kerla_api::driver::block::BlockDriver;
+use kerla_api::{info, warn};
+use kerla_api::mm::{alloc_pages, AllocPageFlags};
+use kerla_api::arch::PAGE_SIZE;
+
+use kerla_api::sync::SpinLock;
+use kerla_utils::alignment::align_up;
+
+use memoffset::offset_of;
+
+use alloc::sync::Arc;
+
+use virtio::device::{IsrStatus, VirtQueue, Virtio, VirtqDescBuffer, VirtqUsedChain};
+
+use virtio::transports::VirtioTransport;
+use virtio::transports::virtio_mmio::VirtioMmio;
+use virtio::transports::virtio_pci::{VirtioAttachError, VirtioPci};
+
+
+const VIRTIO_BLK_F_SIZE: u64 = 1 << 6;
+const VIRTIO_REQUEST_QUEUE: u16 = 0;
+const MAX_BLK_SIZE: usize = 512;
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
+struct VirtioBlkConfig {
+    capacity: u64,
+    size_max: u32,
+    seg_max: u32,
+    cylinders: u16,
+    heads: u8,
+    sectors: u8,
+    blk_size: u32,
+    physical_block_exp: u8,
+    alignment_offset: u8,
+    min_io_size: u16,
+    opt_io_size: u32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
+struct VirtioBlkRequest {
+    type_: u32,
+    reserved: u32,
+    sector: u64,
+}
+
+
+
+#[repr(u32)]
+enum RequestType {
+    Read = 0,
+    Write = 1
+}
+
+
+pub struct VirtioBlk {
+    virtio: Virtio,
```

Comment:
Please confirm it always have to claim ownership of underlying `Virtio` instance?

Reply:
I'm not sure what you mean by claim ownership of "it".